argon2_remote_support.go

```diff
@@ -0,0 +1,460 @@
+			ErrorType:   Argon2RemoteErrorInvalidMode,
+			ErrorString: fmt.Sprintf("invalid mode: %q", string(input.Mode)),
+		}
+	}
+
+	costParams := &Argon2CostParams{
+		Time:      input.Time,
+		MemoryKiB: input.MemoryKiB,
+		Threads:   input.Threads,
+	}
+	if costParams.Time == 0 {
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidTimeCost,
+			ErrorString: "invalid time cost: cannot be zero",
+		}
+	}
+	if costParams.Threads == 0 {
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidThreads,
+			ErrorString: "invalid threads: cannot be zero",
+		}
+	}
+
+	switch input.Command {
+	case Argon2RemoteCommandDerive:
+		key, err := InProcessArgon2KDF().Derive(input.Passphrase, input.Salt, input.Mode, costParams, input.Keylen)
+		if err != nil {
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpected,
+				ErrorString: fmt.Sprintf("cannot run derive command: %v", err),
+			}
+
+		}
+		return &Argon2RemoteOutput{
+			Command: input.Command,
+			Key:     key,
+		}
+	case Argon2RemoteCommandTime:
+		if len(input.Passphrase) > 0 {
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpectedInput,
+				ErrorString: "cannot supply passphrase for \"time\" command",
+			}
+		}
+		if len(input.Salt) > 0 {
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpectedInput,
+				ErrorString: "cannot supply salt for \"time\" command",
+			}
+		}
+		if input.Keylen > 0 {
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpectedInput,
+				ErrorString: "cannot supply keylen for \"time\" command",
+			}
+		}
+
+		duration, err := InProcessArgon2KDF().Time(input.Mode, costParams)
+		if err != nil {
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpected,
+				ErrorString: fmt.Sprintf("cannot run time command: %v", err),
+			}
+		}
+		return &Argon2RemoteOutput{
+			Command:  input.Command,
+			Duration: duration,
+		}
+	default:
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidCommand,
+			ErrorString: fmt.Sprintf("invalid command: %q", string(input.Command)),
+		}
+	}
+}
+
+// WaitAndRunArgon2RequestInRemoteProcess waits for a [Argon2RemoteInput] request on the
+// supplied io.Reader before running it and sending a [Argon2RemoteOutput] response back via
+// the supplied io.Writer. These will generally be connected to the process's os.Stdin and
+// os.Stdout when using - certainly when using [NewRemoteArgon2KDF] on the parent side.
+// This function can only be called once in a process. Subsequent calls in the same
+// process will result in an error response being returned.
+func WaitAndRunArgon2RequestInRemoteProcess(in io.Reader, out io.Writer) error {
```

Comment:
this function code does not seem to be exercised

Reply:
It gets executed in a sub-process (`cmd/run_argon2`) by any test that makes use of `NewOutOfProcessArgon2KDF`, which might be why it doesn't appear in the test coverage.